File: Project2_Multithreaded_File_Management_System/file_manager.py
import threading
import time
from file_operations import FileManager
from logger import log_operation

def perform_operations(file_manager):
    # Example operations
    file_manager.write_file('example.txt', 'Hello World!')
    log_operation('Write', 'example.txt')

    time.sleep(1)  # Simulate some delay

    file_manager.read_file('example.txt')
    log_operation('Read', 'example.txt')

    file_manager.rename_file('example.txt', 'example_renamed.txt')
    log_operation('Rename', 'example.txt')

    file_manager.display_metadata('example_renamed.txt')

    file_manager.copy_file('example_renamed.txt', 'example_copy.txt')
    log_operation('Copy', 'example_renamed.txt')

    file_manager.delete_file('example_copy.txt')
    log_operation('Delete', 'example_copy.txt')

if __name__ == "__main__":
    file_manager = FileManager()
    
    threads = []
    for i in range(5):  # Creating multiple threads for demonstration
        thread = threading.Thread(target=perform_operations, args=(file_manager,))
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

File: Project2_Multithreaded_File_Management_System/file_manager.c
// file_manager.c
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include "file_operations.c"
#include <unistd.h>
#include <wait.h>

sem_t file_semaphore;
sem_t read_lock;

int read_count = 0;

void* perform_operations(void* args) {

    int choice = (*(int*)args);

    const char* filename = "example.txt";

    if (choice == 2){
        // Writing to the file
        sem_wait(&file_semaphore);  // Enter critical section
        write_file(filename);
        sleep(1);
        sem_post(&file_semaphore);  // Leave critical section
    }
    
    else if (choice == 1){
        // Reading from the file
        sem_wait(&read_lock);
        read_count++;
        if (read_count == 1){
            sem_wait(&file_semaphore);
        }
        sem_post(&read_lock);
        
        read_file(filename);

        read_count--;
        if (read_count == 0){
            sem_post(&file_semaphore);
        }
        sem_post(&read_lock);
    }

    else if (choice == 4){
        // Renaming the file
        sem_wait(&file_semaphore);
        rename_file();
        sem_post(&file_semaphore);
    }
    
    else if (choice == 6){
        display_metadata(renamed_filename);
    }
    
    else if (choice == 3){
        // Copying the file
        sem_wait(&file_semaphore);
        copy_file(filename);
        sem_post(&file_semaphore);
    }
    
    else if (choice == 5){
        // Deleting the copied file
        sem_wait(&file_semaphore);
        delete_file(copy_filename);
        sem_post(&file_semaphore);
    }

    return NULL;
}
